spinoso-string/src/enc/utf8/mod.rs

```diff
@@ -804,6 +804,98 @@ impl Utf8String {
     }
 }
 
+// Index
+impl Utf8String {
+    #[inline]
+    #[must_use]
+    pub fn index(&self, needle: &[u8], offset: usize) -> Option<usize> {
+        // Decode needle
+        let needle_chars = match must_decode_utf8(needle) {
+            Some(chars) => chars,
+            // Needle containing any invalid UTF-8 should never match in MRI
```

Comment:
According to `irb`:
<img width="1457" alt="image" src="https://github.com/artichoke/artichoke/assets/12410942/3be0ad1e-695f-40d8-86f8-e0cdf773c16a">

Reply:
sweet. I played with this a bit too:

```console
$ irb
[3.2.2] > s = "abc\xFF"
=> "abc\xFF"
[3.2.2] > s.encoding
=> #<Encoding:UTF-8>
[3.2.2] > s.index s
=> nil
[3.2.2] > s.rindex s
=> nil
[3.2.2] > s.index "abc"
=> 0
[3.2.2] > s.rindex "abc"
=> 0
[3.2.2] > s.index "bc\xFF"
=> nil
[3.2.2] > s.rindex "bc\xFF"
=> nil
[3.2.2] > s.index "bc"
=> 1
[3.2.2] > s.rindex "bc"
=> 1
```